src/array/dictionary/mod.rs

```diff
@@ -119,7 +119,9 @@ impl<K: DictionaryKey> DictionaryArray<K> {
     ) -> Result<Self, Error> {
         check_data_type(K::KEY_TYPE, &data_type, values.data_type())?;
 
-        check_indexes(keys.values(), values.len())?;
+        if keys.null_count() != keys.len() {
+            check_indexes(keys.values(), values.len())?;
```

Comment:
shouldln't we check that the valid keys are in bound?

Reply:
Oh, my bad. This should be `keys.null_count() == keys.len()`, meaning that all key values are masked out. 

This is for instance the keys when we create `DictionaryArray([None, None])`. This keys will be `[0, 0]`, but the length of values will be `0`, so it would not suffice the index check.

Reply:
Agree - I mis-read the condition on the if. Thanks @ritchie46 ! 🙇